HIVE-28800: Improve LongColumnStatsAggregator by preventing the overestimation of NDV

ngsg opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
This patch implements a sanity-check in `LongColumnStatsAggregator`. The added check ensures that the estimated NDV is always less than or equal to the maximum number of integers within the estimated range.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
If the estimated range of a long type column is `[A, B]`, then the maximum number of distinct values in the column should be `B - A + 1` (assuming no nulls in the column). Therefore, any integer that is larger than `B - A + 1` cannot be the estimated value of NDV.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### Is the change a dependency upgrade?
<!--
If yes, please attach a file with output from mvn dependency:tree to validate a complete upgrade of dependency.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
The patch is tested manually by printing debug messages in bitvector-enabled HMS. Also, I checked that the existing unit tests contains invalid expected outputs.

deniskuzZ commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/LongColumnStatsAggregator.java:
```diff
@@ -242,6 +242,24 @@ public ColumnStatisticsObj aggregate(List<ColStatsObjWithSourceInfo> colStatsWit
     statsObj.setStatsData(columnStatisticsData);
+
+    // NDV cannot exceed the number of integers within the given range.
+    long numIntInRange = columnStatisticsData.getLongStats().getHighValue()
```
should we fix this as well
````
densityAvgSum += ((double) (newData.getHighValue() - newData.getLowValue())) / newData.getNumDVs();
````
since we could try to divide by 0
````
double densityAvg = densityAvgSum / partNames.size();
estimation = (long) ((aggregateData.getHighValue() - aggregateData.getLowValue()) / densityAvg);
````
````

ngsg replied:
```
densityAvgSum += ((double) (newData.getHighValue() - newData.getLowValue())) / newData.getNumDVs();
```
There might be an invariant such as columnStats being present implies `NDV != 0`, but it's safer to explicitly guard against division by zero. I'll update the code accordingly.

---

```
double densityAvg = densityAvgSum / partNames.size();
```
If `partNames.size()` is 0, we should return an empty ColumnStats or null, as there's nothing to aggregate. I'll check the convention and adjust the code to handle this case properly.

---

```
estimation = (long) ((aggregateData.getHighValue() - aggregateData.getLowValue()) / densityAvg);
```
It seems that `densityAvg` is 0 in some cases. I'll fix it as well.

deniskuzZ commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/LongColumnStatsAggregator.java:
```diff
@@ -242,6 +242,24 @@ public ColumnStatisticsObj aggregate(List<ColStatsObjWithSourceInfo> colStatsWit
 
     statsObj.setStatsData(columnStatisticsData);
+
+    // NDV cannot exceed the number of integers within the given range.
```
should we move the NDV correction logic inside of `extrapolate`?

ngsg replied:
We may use a probabilistic data structure instead of an extrapolation method, so we could not embed the NDV correction logic inside `extrapolate`. I'll extract the correction logic into a separate method.

deniskuzZ commented:
hi @ngsg, not sure if you saw the comments

ngsg commented:
@deniskuzZ I've addressed your review and updated the patch accordingly. For the division by zero concerns, I didn't modify the division by `partNames.size()` and `ndvEstimator.estimateNumDistinctValues()` because, if I understand them correctly, they are always positive. Could you please review the changes? Thank you.

deniskuzZ commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/LongColumnStatsAggregator.java:
```diff
@@ -129,7 +129,16 @@ public ColumnStatisticsObj aggregate(List<ColStatsObjWithSourceInfo> colStatsWit
           double densityAvg = densityAvgSum / partNames.size();
-          estimation = (long) ((aggregateData.getHighValue() - aggregateData.getLowValue()) / densityAvg);
+          if (densityAvg == 0) {
+            // This means each partition has exactly one value. (highValue == lowValue and NDV = 1).
```
👍

deniskuzZ commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/LongColumnStatsAggregator.java:
```diff
@@ -192,7 +207,13 @@ public ColumnStatisticsObj aggregate(List<ColStatsObjWithSourceInfo> colStatsWit
               adjustedStatsMap.put(pseudoPartName.toString(), csd);
               if (useDensityFunctionForNDVEstimation) {
-                densityAvgSum += ((double) (aggregateData.getHighValue() - aggregateData.getLowValue())) / aggregateData.getNumDVs();
+                if (aggregateData.getNumDVs() == 0) {
```
could it be `0` here as well?
https://github.com/apache/hive/pull/5674/files#diff-004a5c809d016b31028b52e5a5e591d30b446d974b0d22e386c35acfee845d59L227

deniskuzZ commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/LongColumnStatsAggregator.java:
```diff
@@ -192,7 +207,13 @@ public ColumnStatisticsObj aggregate(List<ColStatsObjWithSourceInfo> colStatsWit
               adjustedStatsMap.put(pseudoPartName.toString(), csd);
               if (useDensityFunctionForNDVEstimation) {
-                densityAvgSum += ((double) (aggregateData.getHighValue() - aggregateData.getLowValue())) / aggregateData.getNumDVs();
+                if (aggregateData.getNumDVs() == 0) {
```
could it be `0` after `ndvEstimator.estimateNumDistinctValues()`? if yes, isn't it the same here?
https://github.com/apache/hive/pull/5674/files#diff-004a5c809d016b31028b52e5a5e591d30b446d974b0d22e386c35acfee845d59R248

ngsg replied:
You're right; according to my previous comment, this should be also nonzero. Thank you for pointing out my mistake. I'll remove is-zero test here.

deniskuzZ commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/LongColumnStatsAggregator.java:
```diff
@@ -163,7 +172,13 @@ public ColumnStatisticsObj aggregate(List<ColStatsObjWithSourceInfo> colStatsWit
           LongColumnStatsData newData = cso.getStatsData().getLongStats();
           if (useDensityFunctionForNDVEstimation) {
-            densityAvgSum += ((double) (newData.getHighValue() - newData.getLowValue())) / newData.getNumDVs();
+            if (newData.getNumDVs() == 0) {
```
could it be `0` here as well?
https://github.com/apache/hive/pull/5674/files#diff-004a5c809d016b31028b52e5a5e591d30b446d974b0d22e386c35acfee845d59R105

ngsg replied:
I write a test to validate that `newData.getNumDVs() == 0`. I check the test using LlapLocalDriver with an additional modification to `data/conf/llap/hivemetastore-site.xml`, specifically setting `metastore.stats.ndv.densityfunction` to `true`.

However, I think is-zero test is unnecessary at this point, as JVM does not throw an error even when the divisor is 0 (cf. p672 of the [Java Language Specification](https://docs.oracle.com/javase/specs/jls/se24/jls24.pdf)).

If NDV is 0, this likely indicates that there are no rows in the given partition. In such cases, if high and low values are present(which is a weird situation), the density should be set to infinity. I don’t think we need to address this situation explicitly, as it should not occur under normal circumstances.

For documentation purposes, I attach the test I wrote below:
```
create table x_inventory (inv_quantity_on_hand int) partitioned by (inv_date_sk bigint);
insert into table x_inventory (inv_quantity_on_hand, inv_date_sk) values (0, 2000), (1, 2001), (2, 2002);

set hive.stats.ndv.algo=fm;
analyze table x_inventory compute statistics for columns;

alter table x_inventory partition (inv_date_sk = 2002) update statistics for column inv_quantity_on_hand set('numDVs'='2','numNulls'='0');
alter table x_inventory partition (inv_date_sk = 2001) update statistics for column inv_quantity_on_hand set('numDVs'='1','numNulls'='0');
alter table x_inventory partition (inv_date_sk = 2000) update statistics for column inv_quantity_on_hand set('numDVs'='0','numNulls'='0');

-- at this point, stats uses FMSketch.

set hive.stats.ndv.algo=hll;
insert into table x_inventory (inv_quantity_on_hand, inv_date_sk) values (3, 2003);

-- so there are FMSketch and HyperLogLog.

set hive.stats.column.autogather=false;
insert into table x_inventory (inv_quantity_on_hand, inv_date_sk) values (4, 2004);
-- Finally, this ensures that the colstats is incomplete and not mergable.

desc formatted x_inventory partition (inv_date_sk=2002) inv_quantity_on_hand;
desc formatted x_inventory partition (inv_date_sk=2003) inv_quantity_on_hand;
desc formatted x_inventory partition (inv_date_sk=2004) inv_quantity_on_hand;

-- This requests aggregated column stats of inv_quantity_on_hand.
explain select count(distinct inv_quantity_on_hand) from x_inventory;

```

deniskuzZ commented:
@ngsg, added a few comments, please. take a look. I think `0` check is only needed at line #132

ngsg commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/LongColumnStatsAggregator.java:
```diff
@@ -192,7 +207,13 @@ public ColumnStatisticsObj aggregate(List<ColStatsObjWithSourceInfo> colStatsWit
+                  // If NDV is not present, assume that the density is 0.5.
+                  densityAvgSum += 0.5;
+                } else {
+                  double range = aggregateData.getHighValue() - aggregateData.getLowValue();
```
I removed `range` variable and reverted the code to original version.

ngsg commented:
@deniskuzZ , I have checked your comments and agree that we only need to guard division-by-zero issue at `#132`, as you mentioned. I have added a comment for each case and updated the patch. Could you please confirm them? Thank you.

deniskuzZ reviewed (approved):
LGTM +1, pending tests

deniskuzZ commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/LongColumnStatsAggregator.java:
```diff
@@ -343,4 +356,23 @@ public void extrapolate(ColumnStatisticsData extrapolateData, int numParts,
+
+  private void adjustNDV(ColumnStatisticsData columnStatisticsData) {
+    // NDV cannot exceed the number of integers within the given range.
+    long numIntInRange = columnStatisticsData.getLongStats().getHighValue()
```
@ngsg , could we please rename var to `estimation`

ngsg replied:
Sure, I renamed it to `estimation`.

deniskuzZ commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/LongColumnStatsAggregator.java:
```diff
@@ -343,4 +356,23 @@ public void extrapolate(ColumnStatisticsData extrapolateData, int numParts,
     extrapolateLongData.setNumDVs(ndv);
   }
+
+  private void adjustNDV(ColumnStatisticsData columnStatisticsData) {
```
please change param to `LongColumnStatsData columnStats` <-- `columnStatisticsData.getLongStats()`

ngsg replied:
I changed the parameter to `LongColumnStatsData columnStats`.
